File: swift/Sources/MDoc/Hpke.swift
import CborCose
import Crypto
import Foundation

/// HPKE (RFC 9180) — base mode, cipher suite DHKEM(P-256, HKDF-SHA256) / HKDF-SHA256 / AES-128-GCM.
/// The suite the ISO/IEC 18013-7:2025 Annex C Digital Credentials API uses to encrypt the mdoc
/// `DeviceResponse` to the verifier's ephemeral `recipientPublicKey`. Single-shot sealing only.
public enum Hpke {
    private static let kemId = 0x0010, kdfId = 0x0001, aeadId = 0x0001
    private static let nSecret = 32, nk = 16, nn = 12, nh = 32

    /// The `enc` (encapsulated ephemeral public key, uncompressed) and the AEAD `ciphertext` (incl. tag).
    public struct Sealed { public let enc: [UInt8]; public let ciphertext: [UInt8] }

    /// An HPKE ephemeral P-256 key pair; random in production, injectable for RFC test vectors.
    public struct Ephemeral {
        let privateKey: P256.KeyAgreement.PrivateKey
        let publicUncompressed: [UInt8] // 0x04 || X || Y
        public static func random() -> Ephemeral {
            let k = P256.KeyAgreement.PrivateKey()
            return Ephemeral(privateKey: k, publicUncompressed: [UInt8](k.publicKey.x963Representation))
        }
        static func of(scalar: [UInt8], publicUncompressed: [UInt8]) throws -> Ephemeral {
            Ephemeral(privateKey: try P256.KeyAgreement.PrivateKey(rawRepresentation: Data(scalar)), publicUncompressed: publicUncompressed)
        }
    }

    /// Seals `plaintext` to `recipient` with the given HPKE `info`/`aad`. `ephemeral` is injectable for test vectors.
    public static func sealBaseP256(recipient: EcPublicKey, info: [UInt8], aad: [UInt8], plaintext: [UInt8],
                                    ephemeral: Ephemeral = .random()) throws -> Sealed {
        let recipientKa = try P256.KeyAgreement.PublicKey(x963Representation: Data([0x04] + pad(recipient.x) + pad(recipient.y)))
        let dh = try ephemeral.privateKey.sharedSecretFromKeyAgreement(with: recipientKa).withUnsafeBytes { [UInt8]($0) }
        let enc = ephemeral.publicUncompressed
        let kemContext = enc + [UInt8](recipientKa.x963Representation) // enc || pkRm (uncompressed)

        let sharedSecret = extractAndExpand(kemSuiteId(), dh, kemContext, nSecret)
        let pskIdHash = labeledExtract(hpkeSuiteId(), nil, "psk_id_hash", [])
        let infoHash = labeledExtract(hpkeSuiteId(), nil, "info_hash", info)
        let ksContext = [0] + pskIdHash + infoHash // base mode
        let secret = labeledExtract(hpkeSuiteId(), sharedSecret, "secret", [])
        let key = labeledExpand(hpkeSuiteId(), secret, "key", ksContext, nk)
        let baseNonce = labeledExpand(hpkeSuiteId(), secret, "base_nonce", ksContext, nn)

        // Single-shot: sequence number 0, so the per-message nonce equals base_nonce.
        let box = try AES.GCM.seal(Data(plaintext), using: SymmetricKey(data: Data(key)),
                                   nonce: try AES.GCM.Nonce(data: Data(baseNonce)), authenticating: Data(aad))
        return Sealed(enc: enc, ciphertext: [UInt8](box.ciphertext) + [UInt8](box.tag))
    }

    // MARK: - RFC 9180 §4 labeled KDF

    private static func kemSuiteId() -> [UInt8] { [UInt8]("KEM".utf8) + i2osp(kemId, 2) }
    private static func hpkeSuiteId() -> [UInt8] { [UInt8]("HPKE".utf8) + i2osp(kemId, 2) + i2osp(kdfId, 2) + i2osp(aeadId, 2) }

    private static func labeledExtract(_ suiteId: [UInt8], _ salt: [UInt8]?, _ label: String, _ ikm: [UInt8]) -> [UInt8] {
        hkdfExtract(salt, [UInt8]("HPKE-v1".utf8) + suiteId + [UInt8](label.utf8) + ikm)
    }

    private static func labeledExpand(_ suiteId: [UInt8], _ prk: [UInt8], _ label: String, _ info: [UInt8], _ length: Int) -> [UInt8] {
        hkdfExpand(prk, i2osp(length, 2) + [UInt8]("HPKE-v1".utf8) + suiteId + [UInt8](label.utf8) + info, length)
    }

    private static func extractAndExpand(_ suiteId: [UInt8], _ dh: [UInt8], _ kemContext: [UInt8], _ length: Int) -> [UInt8] {
        let eaePrk = labeledExtract(suiteId, nil, "eae_prk", dh)
        return labeledExpand(suiteId, eaePrk, "shared_secret", kemContext, length)
    }

    // MARK: - HKDF-SHA256 (RFC 5869)

    private static func hkdfExtract(_ salt: [UInt8]?, _ ikm: [UInt8]) -> [UInt8] { hmac(salt ?? [UInt8](repeating: 0, count: nh), ikm) }

    private static func hkdfExpand(_ prk: [UInt8], _ info: [UInt8], _ length: Int) -> [UInt8] {
        var out = [UInt8](), t = [UInt8](), i = 1
        while out.count < length {
            t = hmac(prk, t + info + [UInt8(i)])
            out += t
            i += 1
        }
        return Array(out.prefix(length))
    }

    private static func hmac(_ key: [UInt8], _ data: [UInt8]) -> [UInt8] {
        let k = SymmetricKey(data: key.isEmpty ? [UInt8](repeating: 0, count: nh) : key)
        return [UInt8](HMAC<SHA256>.authenticationCode(for: Data(data), using: k))
    }

    // MARK: - helpers

    private static func i2osp(_ value: Int, _ length: Int) -> [UInt8] {
        var out = [UInt8](repeating: 0, count: length), v = value
        for i in stride(from: length - 1, through: 0, by: -1) { out[i] = UInt8(v & 0xFF); v >>= 8 }
        return out
    }

    private static func pad(_ b: [UInt8]) -> [UInt8] {
        b.count >= 32 ? Array(b.suffix(32)) : [UInt8](repeating: 0, count: 32 - b.count) + b
    }
}

File: kotlin/mdoc/src/main/kotlin/com/hopae/eudi/wallet/mdoc/Hpke.kt
package com.hopae.eudi.wallet.mdoc

import com.hopae.eudi.wallet.cbor.cose.EcPublicKey
import com.hopae.eudi.wallet.cbor.cose.Ecdsa
import java.io.ByteArrayOutputStream
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.interfaces.ECPrivateKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPrivateKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * HPKE (RFC 9180) — base mode, cipher suite DHKEM(P-256, HKDF-SHA256) / HKDF-SHA256 / AES-128-GCM.
 * This is the suite the ISO/IEC 18013-7:2025 Annex C Digital Credentials API uses to encrypt the mdoc
 * `DeviceResponse` to the verifier's ephemeral `recipientPublicKey`. Only single-shot sealing is needed.
 */
object Hpke {
    private const val KEM_ID = 0x0010 // DHKEM(P-256, HKDF-SHA256)
    private const val KDF_ID = 0x0001 // HKDF-SHA256
    private const val AEAD_ID = 0x0001 // AES-128-GCM
    private const val NSECRET = 32 // KEM shared secret
    private const val NK = 16 // AES-128 key
    private const val NN = 12 // GCM nonce
    private const val NH = 32 // SHA-256 output

    /** The `enc` (encapsulated ephemeral public key, uncompressed) and the AEAD `ciphertext` (incl. tag). */
    class Sealed(val enc: ByteArray, val ciphertext: ByteArray)

    /** Seals [plaintext] to [recipient] with the given HPKE `info`/`aad`. [ephemeral] is injectable for test vectors. */
    fun sealBaseP256(
        recipient: EcPublicKey,
        info: ByteArray,
        aad: ByteArray,
        plaintext: ByteArray,
        ephemeral: Ephemeral = Ephemeral.random(),
    ): Sealed {
        val recipientPub = Ecdsa.publicKeyOf(recipient) as ECPublicKey
        val dh = KeyAgreement.getInstance("ECDH").run {
            init(ephemeral.privateKey)
            doPhase(recipientPub, true)
            generateSecret() // P-256 ECDH → 32-byte X coordinate
        }
        val enc = serialize(ephemeral.publicPoint)
        val kemContext = enc + serialize(recipientPub.w)

        val sharedSecret = extractAndExpand(kemSuiteId(), dh, kemContext, NSECRET)

        val pskIdHash = labeledExtract(hpkeSuiteId(), null, "psk_id_hash", ByteArray(0))
        val infoHash = labeledExtract(hpkeSuiteId(), null, "info_hash", info)
        val ksContext = byteArrayOf(0) + pskIdHash + infoHash // base mode
        val secret = labeledExtract(hpkeSuiteId(), sharedSecret, "secret", ByteArray(0))
        val key = labeledExpand(hpkeSuiteId(), secret, "key", ksContext, NK)
        val baseNonce = labeledExpand(hpkeSuiteId(), secret, "base_nonce", ksContext, NN)

        // Single-shot: sequence number 0, so the per-message nonce equals base_nonce.
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, baseNonce))
        cipher.updateAAD(aad)
        return Sealed(enc, cipher.doFinal(plaintext))
    }

    // ---- RFC 9180 §4 labeled KDF ----

    private fun kemSuiteId() = "KEM".toByteArray() + i2osp(KEM_ID, 2)
    private fun hpkeSuiteId() = "HPKE".toByteArray() + i2osp(KEM_ID, 2) + i2osp(KDF_ID, 2) + i2osp(AEAD_ID, 2)

    private fun labeledExtract(suiteId: ByteArray, salt: ByteArray?, label: String, ikm: ByteArray): ByteArray {
        val labeledIkm = "HPKE-v1".toByteArray() + suiteId + label.toByteArray() + ikm
        return hkdfExtract(salt, labeledIkm)
    }

    private fun labeledExpand(suiteId: ByteArray, prk: ByteArray, label: String, info: ByteArray, length: Int): ByteArray {
        val labeledInfo = i2osp(length, 2) + "HPKE-v1".toByteArray() + suiteId + label.toByteArray() + info
        return hkdfExpand(prk, labeledInfo, length)
    }

    private fun extractAndExpand(suiteId: ByteArray, dh: ByteArray, kemContext: ByteArray, length: Int): ByteArray {
        val eaePrk = labeledExtract(suiteId, null, "eae_prk", dh)
        return labeledExpand(suiteId, eaePrk, "shared_secret", kemContext, length)
    }

    // ---- HKDF-SHA256 (RFC 5869) ----

    private fun hkdfExtract(salt: ByteArray?, ikm: ByteArray): ByteArray = hmac(salt ?: ByteArray(NH), ikm)

    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val out = ByteArrayOutputStream()
        var t = ByteArray(0)
        var i = 1
        while (out.size() < length) {
            t = hmac(prk, t + info + byteArrayOf(i.toByte()))
            out.write(t)
            i++
        }
        return out.toByteArray().copyOf(length)
    }

    private fun hmac(key: ByteArray, data: ByteArray): ByteArray = Mac.getInstance("HmacSHA256").run {
        // JCA rejects an all-zero-length key; an all-zero 32-byte key is the RFC 5869 default salt anyway.
        init(SecretKeySpec(if (key.isEmpty()) ByteArray(NH) else key, "HmacSHA256"))
        doFinal(data)
    }

    // ---- P-256 point (de)serialization ----

    private fun serialize(point: ECPoint): ByteArray =
        byteArrayOf(0x04) + fixed(point.affineX, 32) + fixed(point.affineY, 32)

    private fun fixed(v: BigInteger, size: Int): ByteArray {
        val s = v.toByteArray().let { if (it.size > size) it.copyOfRange(it.size - size, it.size) else it }
        return ByteArray(size - s.size) + s
    }

    /** An HPKE ephemeral P-256 key pair; random in production, injectable for RFC test vectors. */
    class Ephemeral private constructor(val privateKey: ECPrivateKey, val publicPoint: ECPoint) {
        companion object {
            fun random(): Ephemeral {
                val kp = KeyPairGenerator.getInstance("EC").apply { initialize(ECGenParameterSpec("secp256r1")) }.generateKeyPair()
                return Ephemeral(kp.private as ECPrivateKey, (kp.public as ECPublicKey).w)
            }

            /** For test vectors: an ephemeral built from a raw private scalar + its public point. */
            internal fun of(scalar: ByteArray, publicPoint: ECPoint): Ephemeral {
                val params = AlgorithmParameters.getInstance("EC").apply { init(ECGenParameterSpec("secp256r1")) }
                    .getParameterSpec(ECParameterSpec::class.java)
                val priv = KeyFactory.getInstance("EC").generatePrivate(ECPrivateKeySpec(BigInteger(1, scalar), params)) as ECPrivateKey
                return Ephemeral(priv, publicPoint)
            }
        }
    }

    private fun i2osp(value: Int, length: Int): ByteArray {
        val out = ByteArray(length)
        var v = value
        for (i in length - 1 downTo 0) { out[i] = (v and 0xFF).toByte(); v = v ushr 8 }
        return out
    }
}
